File: ios/AdjutantKit/Sources/AdjutantKit/Models/Session.swift
import Foundation

/// A managed tmux session tracked by the Session Bridge.
/// Mirrors the backend `ManagedSession` type from `types/session.ts`.
public struct ManagedSession: Codable, Identifiable, Equatable, Hashable {
    public let id: String
    public let name: String
    public let tmuxSession: String
    public let tmuxPane: String
    public let projectPath: String
    public let mode: SessionMode
    public let status: SessionStatus
    public let workspaceType: WorkspaceType
    public let connectedClients: [String]
    public let pipeActive: Bool
    public let createdAt: String
    public let lastActivity: String

    public init(
        id: String,
        name: String,
        tmuxSession: String,
        tmuxPane: String,
        projectPath: String,
        mode: SessionMode,
        status: SessionStatus,
        workspaceType: WorkspaceType,
        connectedClients: [String],
        pipeActive: Bool,
        createdAt: String,
        lastActivity: String
    ) {
        self.id = id
        self.name = name
        self.tmuxSession = tmuxSession
        self.tmuxPane = tmuxPane
        self.projectPath = projectPath
        self.mode = mode
        self.status = status
        self.workspaceType = workspaceType
        self.connectedClients = connectedClients
        self.pipeActive = pipeActive
        self.createdAt = createdAt
        self.lastActivity = lastActivity
    }
}

/// Session operating mode
public enum SessionMode: String, CaseIterable {
    case swarm
    case gastown
}

extension SessionMode: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        // Map legacy "standalone" to .swarm
        if rawValue == "standalone" {
            self = .swarm
            return
        }
        guard let mode = SessionMode(rawValue: rawValue) else {
            // Default to .swarm for unknown values
            self = .swarm
            return
        }
        self = mode
    }
}

/// Session operational status
public enum SessionStatus: String, CaseIterable {
    case idle
    case working
    case waitingPermission = "waiting_permission"
    case offline
}

extension SessionStatus: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        // Fall back to .offline for unknown statuses
        self = SessionStatus(rawValue: rawValue) ?? .offline
    }
}

/// Workspace type for the session
public enum WorkspaceType: String, CaseIterable {
    case primary
    case worktree
    case copy
}

extension WorkspaceType: Codable {
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawValue = try container.decode(String.self)
        // Fall back to .primary for unknown workspace types
        self = WorkspaceType(rawValue: rawValue) ?? .primary
    }
}

/// Request body for creating a new session
public struct CreateSessionRequest: Encodable {
    public let name: String?
    public let projectPath: String
    public let mode: String
    public let workspaceType: String?

    public init(
        name: String? = nil,
        projectPath: String,
        mode: String = "swarm",
        workspaceType: String? = nil
    ) {
        self.name = name
        self.projectPath = projectPath
        self.mode = mode
        self.workspaceType = workspaceType
    }
}

/// Response from killing a session
public struct KillSessionResponse: Codable, Equatable {
    public let killed: Bool
}

/// Response from session discovery
public struct DiscoverSessionsResponse: Codable, Equatable {
    public let discovered: Int
    public let sessions: [ManagedSession]
}

/// Request body for sending input to a session
public struct SendSessionInputRequest: Encodable {
    public let text: String

    public init(text: String) {
        self.text = text
    }
}

/// Response from sending input to a session
public struct SendSessionInputResponse: Codable, Equatable {
    public let sent: Bool
}

File: backend/src/types/session.ts
/**
 * Session Bridge v2 types.
 *
 * Defines the core data model for managed tmux sessions
 * and the WebSocket v2 protocol message types.
 */

// ============================================================================
// Session Data Model
// ============================================================================

export type SessionMode = "standalone" | "swarm" | "gastown";
export type SessionStatus = "idle" | "working" | "waiting_permission" | "offline";
export type WorkspaceType = "primary" | "worktree" | "copy";

/** A managed tmux session tracked by the Session Bridge. */
export interface ManagedSession {
  id: string;
  name: string;
  tmuxSession: string;
  tmuxPane: string;
  projectPath: string;
  mode: SessionMode;
  status: SessionStatus;
  workspaceType: WorkspaceType;
  connectedClients: string[];
  pipeActive: boolean;
  createdAt: string;
  lastActivity: string;
}

/** Serialized form for file persistence (no Sets, no Dates). */
export interface SessionFile {
  sessions: ManagedSession[];
  savedAt: string;
}

// ============================================================================
// Ring Buffer for Output
// ============================================================================

export interface RingBufferOptions {
  maxLines: number;
}

// ============================================================================
// WebSocket v2 Protocol — Client → Server
// ============================================================================

export interface SessionListRequest {
  type: "session_list";
}

export interface SessionCreateRequest {
  type: "session_create";
  projectPath: string;
  mode: SessionMode;
  name?: string;
  workspaceType?: WorkspaceType;
}

export interface SessionConnectRequest {
  type: "session_connect";
  sessionId: string;
  replay?: boolean;
}

export interface SessionDisconnectRequest {
  type: "session_disconnect";
  sessionId: string;
}

export interface SessionInputRequest {
  type: "session_input";
  sessionId: string;
  text: string;
}

export interface SessionInterruptRequest {
  type: "session_interrupt";
  sessionId: string;
}

export interface SessionKillRequest {
  type: "session_kill";
  sessionId: string;
}

export interface SessionPermissionResponse {
  type: "session_permission";
  sessionId: string;
  requestId: string;
  approved: boolean;
}

export type SessionClientMessage =
  | SessionListRequest
  | SessionCreateRequest
  | SessionConnectRequest
  | SessionDisconnectRequest
  | SessionInputRequest
  | SessionInterruptRequest
  | SessionKillRequest
  | SessionPermissionResponse;

// ============================================================================
// WebSocket v2 Protocol — Server → Client
// ============================================================================

export interface SessionListResponse {
  type: "session_list";
  sessions: ManagedSession[];
}

export interface SessionCreatedEvent {
  type: "session_created";
  session: ManagedSession;
}

export interface SessionOutputEvent {
  type: "session_output";
  sessionId: string;
  events: import("../services/output-parser.js").OutputEvent[];
  raw: string;
}

export interface SessionRawEvent {
  type: "session_raw";
  sessionId: string;
  data: string;
}

export interface SessionStatusEvent {
  type: "session_status";
  sessionId: string;
  status: SessionStatus;
}

export interface SessionEndedEvent {
  type: "session_ended";
  sessionId: string;
  reason: string;
}

export interface SessionErrorEvent {
  type: "session_error";
  sessionId?: string;
  message: string;
}

export type SessionServerMessage =
  | SessionListResponse
  | SessionCreatedEvent
  | SessionOutputEvent
  | SessionRawEvent
  | SessionStatusEvent
  | SessionEndedEvent
  | SessionErrorEvent;
